// jspt/lib/jspt/interpreter.js
/**
 * Javascript Portugol
 * https://github.com/moacir/jspt
 *
 * Copyright (c) 2013 Moacir de Oliveira
 * Licensed under the MIT license.
 */

'use strict';

var AST = require('./ast').ast;

function Interpreter(ast, context) {
    this.ast = ast;
    this.context = context || {};
}

Interpreter.VAR_DEFAULT_VALUES = {
    'inteiro': 0,
    'real': 0.0,
    'literal': '',
    'caractere': '',
    'lógico': 'false'
};

Interpreter.prototype = {
    execute: function () {
        return this.visit(this.ast, this.context);
    },
    visit: function (node, context) {
        var method = 'visit' + node.name + 'Node';
        console.log(method);
        //alert(JSON.stringify(node));
        if (this[method] === undefined) {
            global.terminal.error('Invalid Node: ' + node.name);
            throw new Error('Invalid Node: ' + node.name);
        }

        return this[method].call(this, node, context);
    },
    visitAlgorithmNode: function (node, context) {
        if (node.variables) {
            this.visit(node.variables, context);
        }
        if (node.functions) {
            this.visit(node.functions, context);
        }

        return this.visit(node.body, context);
    },
    visitVariablesBlockNode: function (node, context) {
        var list = node.list || [];

        for (var i=0; i<list.length; i++) {
            this.visit(list[i], context);
        }
    },
    visitVariableDeclarationNode: function (node, context) {
        var defaultValue = Interpreter.VAR_DEFAULT_VALUES[node.type];

        context.setVariable(node.id, node.type, defaultValue);
    },
    visitAssignmentExpressionNode: function (node, context) {
        var ident = this.visit(node.left, context),
            right = this.visit(node.right, context);
        // right.type = ident.type;
        // alert ("IDENT :" + JSON.stringify(ident));
        // alert ("RIGHT :" + JSON.stringify(right));

        if (right.type=='qualquer'){right.type=ident.type;}
        if (ident.type != right.type) {
            global.terminal.error('#1 Erro de Tipagem: Variável do tipo: ' + ident.type + '. Encontrado: ' + right.type);
            throw new Error('#1 Erro de Tipagem: Variável do tipo: ' + ident.type + '. Encontrado: ' + right.type);
        }

        ident.value = this.getValue(right);
    },
    visitFunctionDeclarationListNode: function (node, context) {
        var list = node.list || [];

        for (var i=0; i<list.length; i++) {
            this.visit(list[i], context);
        }
    },
    visitFunctionDeclarationNode: function (node, context) {
        if (context.exists(node.id)) {
            global.terminal.error('Este indentificador já foi definido: ' + node.id);
            throw new Error('Este indentificador já foi definido: ' + node.id);
        }

        node.type = node.type || 'vazio';

        context.setFunction(node.id, node);
    },
    visitFunctionCallNode: function (node, context) {
        if (!context.exists(node.id)) {
            global.terminal.error('Função não definida: ' + node.id);
            throw new Error('Função não definida: ' + node.id);
        }

        var fn = context.getItem(node.id),
            fnContext = {},
            args = [],
            item = null;

        //Check the number of parameters defined and passed

        if (node.args.length != fn.value.params.length) {
            global.terminal.error('Número de parâmetros errado: Esperado: ' + fn.value.params.length + '. Encontrado: ' + node.args.length);
            throw new Error('Número de parâmetros errado: Esperado: ' + fn.value.params.length + '. Encontrado: ' + node.args.length);
        }

        //Set function parameters in the new Context

        fnContext = context.copy();

        for (var i=0; i<node.args.length; i++) {
            args[i] = this.visit(node.args[i], context);
            // args[i].type = "literal"; //CORRIGE O PROBLEMA DA TIPAGEM AO IMPRIMIR - OLD

            item = fn.value.params[i];

            /*
                Uma vez que para as funções nativas (imprima e leia) não há diferenciação de tipagem, não verificamos as tipagens
            */
            // alert(JSON.stringify(item));
            // alert(JSON.stringify(args[i].type) + " - " + JSON.stringify(item.type));

            if (item.type == 'qualquer') {
                var oldType = item.type;
                item.type = args[i].type;
            }

            // alert(JSON.stringify(args[i].type) + " - " + JSON.stringify(item.type));

            // alert(JSON.stringify(args[i]));

            if (item.type != args[i].type) {
                global.terminal.error('#2 Erro de Tipagem: Variável do tipo: ' + item.type + '. Encontrado: ' + args[i].type);
                throw new Error('#2 Erro de Tipagem: Variável do tipo: ' + item.type + '. Encontrado: ' + args[i].type);
            }

            item.type = oldType;
            fnContext.setVariable(item.id, item.type, args[i].value);
        }

        //Interpret Function Body

        var fnNode = fn.value;
        var ret = null;

        if (fnNode.variables) {
            this.visit({name: 'VariablesBlock', list: fnNode.variables}, fnContext);
        }

        if (fnNode.name == 'NativeFunction') {
            /*
            if (global.left) { //PROBLEMA PRINTAR SEM VARIAVEIS SETADAS
                var esperado = this.visit(global.left, context);
                fnNode.type = esperado.type; //altera o tipo do nó para o tipo esperado
            }
            */

            if (fnNode.id == 'leia') {
                ret = this.visit(fnNode.execute(args)); //ONDE É EXECUTADO O LEIA  alert(JSON.stringify(ret));
            } else {
                ret = this.visit(fnNode.execute(args)); //ONDE É EXECUTADO O IMPRIMA
            }

        } else {
            ret = this.visit(fnNode.body, fnContext);
        }

        if (ret.type != fnNode.type) {
            global.terminal.error('#3 Erro de Tipagem: Função retorna tipo errado. Declarado: ' + fnNode.type + '. Encontrado: ' + ret.type);
            throw new Error('#3 Erro de Tipagem: Função retorna tipo errado. Declarado: ' + fnNode.type + '. Encontrado: ' + ret.type);
        }

        return ret;
    },
    visitFunctionStatementNode: function (node, context) {
        this.visit(node.call, context);
    },
    visitReturnStatementNode: function (node, context) {
        return this.visit(node.argument, context);
    },
    visitIfStatementNode: function (node, context) {
        var test = this.getValue(this.visit(node.test, context));

        if (test) {
            return this.visit(node.consequent, context);
        }
        if (node.alternate) {
            return this.visit(node.alternate, context);
        }
    },
    visitWhileStatementNode: function (node, context) {
        var ret = null;
        while (this.getValue(this.visit(node.test, context))) {
            if ((ret = this.visit(node.body, context).value) !== undefined) {
                return ret;
            }
        }
    },
    visitForStatementNode: function (node, context) {
        var index = this.visit(node.variable, context),
            start = this.visit(node.start, context),
            end = this.visit(node.end, context),
            update = 1,
            ret = null;

        if (node.update) {
            update = this.getValue(this.visit(node.update, context));
        }

        start = this.getValue(start);
        end = this.getValue(end);

        for (index.value = start; (start <= end) ? index.value <= end : index.value >= end; index.value += update) {
            if ((ret = this.visit(node.body, context).value) !== undefined) {
                return ret;
            }
        }
        context.setFunction(node.value.name, node); //TODO
    },
    visitUnaryExpressionNode: function (node, context) {
        var argument = this.visit(node.argument, context);

        argument = this.getValue(argument);

        switch (node.operator) {
            case '~':
            case 'não':
                return !argument;
            case '+':
                return +argument;
            case '-':
                return -argument;
        }
        global.terminal.error('Operador unário inválido: ' + node.operator);
        throw new Error('Operador unário inválido: ' + node.operator);
    },
    visitBinaryExpressionNode: function (node, context) {
        var left = this.getValue(this.visit(node.left, context));
        var right = this.getValue(this.visit(node.right, context));

        if (node.operator == '=') {
            return left == right;
        } else if (node.operator == '<>') {
            return left != right;
        } else if (node.operator == 'e' || node.operator == '&&') {
            return left && right;
        } else if (node.operator == 'ou' || node.operator == '||') {
            return left || right;
        } else if (node.operator == '^') {
            return left ^ right;
        } else if (node.operator == '&') {
            return left & right;
        } else if (node.operator == '|') {
            return left | right;
        } else if (node.operator == '>') {
            return left > right;
        } else if (node.operator == '>=') {
            return left >= right;
        } else if (node.operator == '<') {
            return left < right;
        } else if (node.operator == '<=') {
            return left <= right;
        }
        global.terminal.error('Operador Inválido: ' + node.operator);
        throw new Error('Operador Inválido: ' + node.operator);
    },
    visitArithmeticExpressionNode: function (node, context) {
        var left = this.visit(node.left, context);
        var right = this.visit(node.right, context);

        //Type Check RESOLVIDO
        // if (left.type != right.type) {
        //     global.terminal.error('Erro de Tipagem: Expressão entre tipos diferentes: ' + left.type + ' e ' + right.type);
        //     throw new Error('Erro de Tipagem: Expressão entre tipos diferentes: ' + left.type + ' e ' + right.type);
        // }

        //TODO: Handle non numeric types
        //TODO: Handle real/inteiro conversions

        var literal = new AST.LiteralNode(left.type);

        if (node.operator == '+') {
            literal.value = left.value + right.value;
        } else if (node.operator == '-') {
            literal.value = left.value - right.value;
        } else if (node.operator == '*') {
            literal.value = left.value * right.value;
        } else if (node.operator == '/') {
            literal.value = left.value / right.value;
        } else {
            global.terminal.error('Operador Inválido: ' + node.operator);
            throw new Error('Operador Inválido: ' + node.operator);
        }

        literal.raw = literal.value;

        return literal;
    },
    visitLogicalExpressionNode: function (node, context) {
        return this.visitBinaryExpressionNode(node, context);
    },
    visitLiteralNode: function (node, context) {
        if (node.type == 'lógico') {
            node.value =  this.visitBooleanLiteral(node.value);
        } else if (node.type == 'inteiro') {
            node.value = parseInt(node.value, 10);
        } else if (node.type == 'real') {
            node.value = parseFloat(node.value);
        } else if (node.type == 'literal') {
            var str = node.value || '""';
            //node.value = str.substr(1, str.length-2); //remove trailling quotes (original do jspt, usando substr)
            node.value = str.replace(/["]+/g, ''); //remove as aspas usando expressão regular
        }
        //alert(JSON.stringify(node));
        return node;
    },
    visitIdentifierNode: function (node, context) {
        if (!context.exists(node.id)) {
            global.terminal.error('Identificador não declarado: ' + node.id);
            throw new Error('Identificador não declarado: ' + node.id);
        }

        return context.getItem(node.id);
    },
    visitBooleanLiteral: function (value) {
        if (value == 'verdadeiro') {
            return true;
        } else if (value == 'falso') {
            return false;
        } else {
            global.terminal.error('Valor lógico inválido: ' + value);
            throw new Error('Valor lógico inválido: ' + value);
        }
    },
    visitExpressionNode: function (node, context) {
        return this.visit(node.expression || node, context);
    },
    visitStatementListNode: function (node, context) {
        var list = node.list || [],
            ret = null;

        for (var i=0; i<list.length; i++) {
            if ((ret = this.visit(list[i], context)) !== undefined) {
                if ((ret.type) !== "vazio"){
                    return ret; //if there's a return value we must break the statement list
                }
            }
        }

        return new AST.LiteralNode('vazio');
    },
    getValue: function (node) {
        if (typeof node.value !== 'undefined') {
            if (node.type == 'lógico') {
                    node.value =  this.visitBooleanLiteral(node.value);
                } else if (node.type == 'inteiro') {
                    node.value = parseInt(node.value, 10);
                } else if (node.type == 'real') {
                    node.value = parseFloat(node.value);
                } else if (node.type == 'literal') {
                    var str = node.value || '""';
                    //node.value = str.substr(1, str.length-2); //remove trailling quotes (original do jspt, usando substr)
                    node.value = str.replace(/["]+/g, ''); //remove as aspas usando expressão regular
                }
            return node.value;
        }

        return node;
    }
};

exports.interpreter = Interpreter;
