// jspt/lib/jspt/native.js
/**
 * Javascript Portugol
 * https://github.com/moacir/jspt
 *
 * Copyright (c) 2013 Moacir de Oliveira
 * Licensed under the MIT license.
 */

'use strict';

var AST = require('./ast').ast;

function NativeFunction(id, type, params, body) {
    console.log("instanciou um native " + id);
    this.name = 'NativeFunction';
    this.id = id;
    this.type = type || 'vazio';
    this.params = params || [];
    this.body = body || function () {};
}

NativeFunction.prototype.getInfo = function () {
    //alert(JSON.stringify(this));
};

NativeFunction.prototype.addParameter = function (id, type) {
    this.params.push(new AST.FunctionParameterNode(id, type));
};

NativeFunction.prototype.setBody = function (fn) {
    this.body = fn;
};

NativeFunction.prototype.execute = function (args) {
    var values = [];
    //alert(this.type + " - " +typeRet);

    for (var i=0; i<args.length; i++) {
        values[i] = args[i].value;
    }

    var ret = this.body.apply(this, values);
    //if (this.type=='qualquer'){this.type=this.params[0].type;}

    return new AST.LiteralNode(this.type, ret, ret);
};

exports.nativeFunction = NativeFunction;
